snuba/utils/streams/batching.py

```diff
@@ -94,170 +239,75 @@ def __init__(
         metrics: MetricsBackend,
         recoverable_errors: Optional[Sequence[Type[ConsumerError]]] = None,
     ) -> None:
-        self.consumer = consumer
-
-        assert isinstance(worker, AbstractBatchWorker)
-        self.worker = worker
+        self.__consumer = consumer
 
-        self.max_batch_size = max_batch_size
-        self.max_batch_time = max_batch_time  # in milliseconds
-        self.__metrics = metrics
+        self.__processor_factory = partial(
+            BatchProcessor, consumer, worker, max_batch_size, max_batch_time, metrics
+        )
 
-        self.shutdown = False
+        self.__processor: Optional[BatchProcessor] = None
```

Comment:
This could be replaced with some sort of state pattern implementation to get rid of all the `is None`/`is not None` checks below if that's cleaner, but ultimately the end result will be the same and there are only two states (have an assignment or not) so I'm not sure that's useful.

Reply:
imo, I would not use a State pattern for this case. It does not seem you would have a common interface between the two states (which is one of the main motivations for the state parttern: a system that takes commands and delegates them to the current state so that none externally needs to know the internals nor the states need to know about each other implementation).
On the other hand doing it in one class would expose the familiar structure of a kafka consumer we are accustomed to.